[jit] Add resolver for 'torch' module

driazati opened the pull request:
This lets you compile builtin functions from C++ without having a dependence on Python

```cpp
auto module = torch::jit::compile(JIT"(
def my_script_method(x, y):
    return torch.relu(x) + y
)");
IValue result = module->run_method("my_script_method", 1, 2);
```

@goldsborough @zdevito @apaszke

Commit:
[jit] Add resolver for 'torch' module

Summary:
This lets you use builtin functions from C++ without having a dependence
on Python

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

goldsborough reviewed (changes requested):
The implementation of `compileFunction` looks great, but it needs more documentation and possibly some usability improvements for running the  function. We want to expose this method to public C++ users so it should be easy to use

Commit:
Add JIT to C++ API

Summary:
* Adds jit.h and jit.cpp to csrc/api
    * Adds torch::jit::compile and torch::jit::run methods to expose
script compiler and interpreter to C++ users
* Add test_api test for JIT API

Test Plan:
`./build/bin/test_api "torch script"`

Reviewers:

Subscribers:

Tasks:

Tags:

zdevito commented on torch/csrc/jit/script/compiler.h:
```diff
@@ -134,6 +149,7 @@ TORCH_API void defineMethodsInModule(
 // same as above but parse the definitions from source
 TORCH_API void defineMethodsInModule(Module & m, const std::string& source, const Resolver& resolver, std::shared_ptr<SugaredValue> self);
 TORCH_API std::shared_ptr<Graph> compileFunction(Def def, const Resolver& resolver);
+TORCH_API std::shared_ptr<Graph> compileFunction(const std::string& source);
```
Just like in python this API should only ever return a jit::script::Module with methods on it. Graphs and their execution are an implementation detail we don't want in the public API. 

For end users the API should probably be something like:
```
std::shared_ptr<jit::script::Module> compile(std::string str) {
   std::shared_ptr<jit::script::Module> foo = new jit::script::Module();
   defineMethodsInModule(...);
   return foo;
}
```

zdevito commented on torch/csrc/jit/test_jit.cpp:
```diff
@@ -864,6 +864,23 @@ void testBlocks(std::ostream & out) {
   out << *g2 << "\n";
 }
 
+static std::vector<IValue> run(std::shared_ptr<Graph> graph, std::vector<IValue> stack) {
```
Directly running the interpreter bypasses all optimizations and may not always work as you would expect. It also does not error checking and would lead to inscrutable errors. By using jit::script::Module's already existing overloaded operators, you will get the right behavior.

apaszke commented on torch/csrc/jit/script/compiler.h:
```diff
@@ -134,6 +149,7 @@ TORCH_API void defineMethodsInModule(
 // same as above but parse the definitions from source
 TORCH_API void defineMethodsInModule(Module & m, const std::string& source, const Resolver& resolver, std::shared_ptr<SugaredValue> self);
 TORCH_API std::shared_ptr<Graph> compileFunction(Def def, const Resolver& resolver);
+TORCH_API std::shared_ptr<Graph> compileFunction(const std::string& source);
```
Well a reason to call it `compileFunction` instead of `compile` is that it can only handle a single function, and it will return a graph instead of a runnable thing.

driazati commented:
The latest commit pulls the functionality out of `torch/jit` and into the C++ API with `compile` and `run` functions.

driazati commented on torch/csrc/api/include/torch/jit.h:
```diff
@@ -0,0 +1,58 @@
+Stack run(std::shared_ptr<script::Module> module, const std::string& method_name, Types&... args) {
+  // Marhsal arguments into stack of IValues
+  Stack stack;
+  std::initializer_list<int>{ (stack.push_back(IValue(args)), 0)... };
```
This seemed a little hacky, is there a better way to handle this case of running a function for all of the var args? Also it's making a copy since the only way to avoid that would be to `std::move` things in but then that would silently take ownership from the caller which could be bad, should we add some way to make an `IValue` with a pointer to the value?

goldsborough replied:
So a `Stack` is just a `std::vector<IValue>`, so I think you could just construct it directly as `Stack stack = {IValue(std::forward<Args>(args))...};`? Let me know if that works

goldsborough commented on torch/csrc/api/include/torch/jit.h:
```diff
@@ -0,0 +1,58 @@
+  std::initializer_list<int>{ (stack.push_back(IValue(args)), 0)... };
+
+  // Interpret graph and run computation
+  module->get_method(method_name).run(stack);
```
We recently added an overload for `operator()` which checks arguments against the schema: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/script/module.h#L63

And then you can also just construct the vector directly when calling the function. So this whole function should maybe be
```
module->get_method(method_name)({IValue(std::forward<Args>(args))...});
```

goldsborough commented on torch/csrc/api/include/torch/jit.h:
```diff
@@ -0,0 +1,58 @@
+/// @param args Arguments to be passed to the method
+/// @return A vector of `IValue`s that contain the results of the method
+template<typename... Types>
+Stack run(std::shared_ptr<script::Module> module, const std::string& method_name, Types&... args) {
```
Right now you're taking all arguments as lvalue references. You want to take them as universal references (`Types&&`) and then use `std::forward` to cast them back to their correct value class. Read https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers if you have time

goldsborough commented on torch/csrc/api/src/jit.cpp:
```diff
@@ -0,0 +1,16 @@
+
+#include <torch/csrc/jit/script/compiler.h>
+#include <torch/csrc/jit/stack.h>
+
```
Please add includes 
```
<memory>
<string>
```

goldsborough commented on torch/csrc/api/include/torch/jit.h:
```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <string>
```
The  header include order should ideally be from most specific to least specific, i.e. your library -> third party libraries -> standard headers, so this would best be

```
#include <torch/csrc/jit/script/module.h>
#include <torch/csrc/jit/script/compiler.h>
#include <torch/csrc/jit/stack.h>

#include <memory>
#include <string>
#include <vector>
```

goldsborough commented on torch/csrc/jit/script/compiler.h:
```diff
@@ -122,8 +122,25 @@ struct TORCH_API BuiltinFunction : public SugaredValue {
 using Resolver = std::function<std::shared_ptr<
     SugaredValue>(const std::string& name, Method& m, const SourceRange& loc)>;
+
+std::shared_ptr<SugaredValue> nativeResolver(const std::string& name, Method& m, const SourceRange& loc);
```
This may need a `TORCH_API`, let's see what windows CI says

goldsborough commented on torch/csrc/api/include/torch/jit.h:
````diff
@@ -0,0 +1,58 @@
+/// auto output = torch::jit::run(module, "test_while", a, b);
+/// ```
+///
+/// @param source A JIT string containing functions that are valid under the
````
I think a "JIT string" isn't really a thing, it's just a string. Also "under the scope of the script compiler" isn't too clear to me. Maybe just "A string containing the script code to compile".

goldsborough commented on test/cpp/api/jit.cpp:
```diff
@@ -0,0 +1,33 @@
+#include <iostream>
```
Order should be

```
catch.hpp

torch/...

std headers
```

Also iostream can be removed

Commit:
Change to universal references

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

zdevito commented on torch/csrc/api/include/torch/jit.h:
```diff
@@ -0,0 +1,57 @@
+/// @return An IValue containing the return value (or values if it is a tuple)
+/// from the method
+template <typename... Types>
+IValue run(
```
If you are adding syntax sugar for running, it should like:
```
auto c = module->run("relu_script", a, b); 
```

It is pretty weird to have this be a method.

driazati replied:
Agreed, fixed in latest push. It's better on the user end but might get messy later on if we're mixing user-end APIs into the internal JIT code. Should we have a wrapper for `Module` for the C++ API instead?

driazati commented:
@pytorchbot retest this please

goldsborough commented on torch/csrc/api/src/jit.cpp:
```diff
@@ -0,0 +1,19 @@
+
+std::shared_ptr<script::Module> compile(const std::string& source) {
+  auto module = std::make_shared<script::Module>();
+  defineMethodsInModule(*module, source, script::nativeResolver, nullptr);
```
Tip: For functions that take many arguments, and especially for values that don't mean anything when read, like `nullptr`, it's good practice to use what we call "LLVM-style comments" by writing `/*name_of_argument=*/value` so that the reader has a better idea of what is going on without having to consult the function declaration

goldsborough commented on torch/csrc/api/include/torch/jit.h:
```diff
@@ -0,0 +1,33 @@
+#include <torch/csrc/jit/script/module.h>
+#include <torch/csrc/jit/stack.h>
+
+#include <string>
```
Also missing `<memory>`

goldsborough commented on torch/csrc/jit/script/compiler.h:
```diff
@@ -122,8 +122,25 @@ struct TORCH_API BuiltinFunction : public SugaredValue {
+    : name(name) {}
+  std::string name;
+
+  virtual std::string kind() const override {
```
`virtual` + `override` is redundant and our linters will complain once I activate them, please only use `override`. I frequently have to codemod the repo to undo this

goldsborough commented on torch/csrc/jit/script/compiler.h:
```diff
@@ -122,8 +122,25 @@ struct TORCH_API BuiltinFunction : public SugaredValue {
+    return "builtin module";
+  }
+
+  virtual std::shared_ptr<SugaredValue> attr(SourceRange loc, Method & m, const std::string& field) override {
```
Same redundant `virtual` + `override` (only need `override`)

goldsborough reviewed (approved):
I made some small suggestions that improve legibility/correctness, but they're nits. You can address them if you want, otherwise feel free to land already

Commit:
Address nits

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:
